src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -47,6 +67,59 @@ public synchronized void delete(Long id) {
         this.productList.stream().filter(product -> product.getId().equals(id))
                 .findFirst()
                 .map(product -> this.productList.remove(product))
-                .orElseThrow(() -> new NoSuchElementException("Product with such ID = " + id + "is not found"));
+                .orElseThrow(() -> new ProductNotFoundException("Product with such ID = " + id + " is not found"));
+    }
+
+    private List<Product> findProductsByQueryParam(String[] query, List<Product> validProducts) {
+        Objects.requireNonNull(validProducts, "Collection should not be null");
+        HashMap<Product, Integer> map = new HashMap<>();
```

Comment:
https://medium.com/@sanchezdale/oop-good-practices-coding-to-the-interface-baea84fd60d3